File: packages/eep-middleware-python/eep_middleware/auth/api_key.py
from __future__ import annotations

from typing import Any, Awaitable, Callable


APIKeyResolver = Callable[[str], Awaitable[dict[str, Any] | None]]


class APIKeyAuthAdapter:
    def __init__(self, resolver: APIKeyResolver) -> None:
        self._resolver = resolver

    async def extract_proofs(self, headers: dict[str, str], query: dict[str, str] | None = None) -> list[dict[str, Any]]:
        api_key = headers.get("x-api-key")
        if not api_key:
            return []

        resolved = await self._resolver(api_key)
        if not resolved:
            return []

        proofs: list[dict[str, Any]] = []
        did_value = resolved.get("did")
        if isinstance(did_value, str) and did_value:
            proofs.append({"type": "identity", "method": "did_verified", "evidence": did_value})

        capabilities = resolved.get("capabilities")
        if isinstance(capabilities, list) and capabilities:
            proofs.append({"type": "capability", "declared_capabilities": capabilities})

        return proofs

File: packages/@eep-dev/middleware/src/auth/api-key.ts
import type { GateProof } from "@eep-dev/gates";
import type { AuthAdapter, IncomingRequest } from "../core/request-handler.js";

export type APIKeyResolver = (apiKey: string) => Promise<{ did?: string; capabilities?: string[] } | null>;

export class APIKeyAuthAdapter implements AuthAdapter {
  constructor(private readonly resolver: APIKeyResolver) {}

  async extractProofs(request: IncomingRequest): Promise<GateProof[]> {
    const apiKey = request.headers["x-api-key"];
    if (!apiKey) {
      return [];
    }

    const resolved = await this.resolver(apiKey);
    if (!resolved) {
      return [];
    }

    const proofs: GateProof[] = [];
    if (resolved.did) {
      proofs.push({ type: "identity", method: "did_verified", evidence: resolved.did });
    }
    if (resolved.capabilities && resolved.capabilities.length > 0) {
      proofs.push({ type: "capability", declared_capabilities: resolved.capabilities });
    }
    return proofs;
  }
}
